### PR TITLE
Update encrypted images

### PR DESCRIPTION
**Pull request by rohang06**
This change adds the update-encrypted-ami option to the brkt_cli.
It allows a user to retrieve a metavisor upgrader AMI and use it
to upgrade an already encrypted AMI. This provides a speed
improvement for users who wish to update their encrypted AMIs.

The user must specify the metavisor updater AMI as an argument.
This requirement will be taken away in a near-future commit.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +48,25 @@ optional arguments:
   --no-validate-ami     Don't validate AMI properties
   --region NAME         AWS region (e.g. us-west-2)
 ```
+```
+$ python brkt update-encrypted-ami --help
+usage: brkt update-encrypted-ami [-h] --updater-ami UPDATER_AMI_ID --region
+                                 REGION [--zone ZONE]
```
Comment (bburtin) [inline README.md]: Let's remove the zone argument, unless it's really necessary.

File: brkt_cli/__init__.py [hunk 2]
```diff
@@ -197,6 +157,147 @@ def main():
...
+    if util.validate_region(region) == 1:
+        print ('Invalid region %s' % region,
+               file=sys.stderr)
+        return 1
+    nonce = util.make_nonce()
+    default_tags = encrypt_ami.get_default_tags(nonce, '')
+    aws_svc = aws_service.AWSService(
+        nonce, None, default_tags=default_tags)
+    aws_svc.connect(region)
+    zones = [str(z.name) for z in aws_svc.conn.get_all_zones()]
```
Comment (bburtin) [inline brkt_cli/__init__.py]: Be careful with aws_svc.conn.  It won't work with unit tests.  AWSService is meant to be an abstraction.

File: README.md (hunk 1 above)
Comment (rohang06) [inline README.md]: Fixed

File: README.md [hunk 3]
```diff
@@ -82,3 +100,27 @@ ami-07c2a262
 
 When the process completes, the new AMI id is written to stdout.  All log
 messages are written to stderr.
+
+## Updating an encrypted AMI
+
+Run **brkt update-encrypted-ami** to update anw encrypted AMI based on an existing
```
Comment (dramdass) [inline README.md]: nit: s/anw/an

**Comment by dramdass** [comment]
Small comment above. Code looks reasonable to me - nice refactors. I'll let Boris give the final approval

File: brkt_cli/update_encrypted_ami.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    guest_volume_mapping = \
+        encrypted_image.block_device_mapping.get('/dev/sda5')
+    if not guest_volume_mapping:
+        return None, None, None, \
+            'Invalid block device mapping: /dev/sda5 not present'
+    snapshot = aws_service.conn.get_all_snapshots(
+        guest_volume_mapping.snapshot_id)[0]
+    # Create new volume off this snapshot, snapshot the volume and
+    # return the snapshot
+    guest_volume = snapshot.create_volume(zone)
```
Comment (dramdass) [inline brkt_cli/update_encrypted_ami.py]: do we need to do this snapshot -> volume -> snapshot logic here? the customer owns the guest root snapshot so we shouldn't need to do this

**Comment by dramdass** [comment]
One more comment here.

File: brkt_cli/__init__.py [hunk 5]
```diff
@@ -197,6 +157,147 @@ def main():
...
+            return 1
+    region = values.region
+    if util.validate_region(region) == 1:
+        print ('Invalid region %s' % region,
+               file=sys.stderr)
+        return 1
+    nonce = util.make_nonce()
+    default_tags = encrypt_ami.get_default_tags(nonce, '')
+    aws_svc = aws_service.AWSService(
+        nonce, None, default_tags=default_tags)
```
Comment (bburtin) [inline brkt_cli/__init__.py]: I just realized that the encryptor_ami argument to the AWSService constructor is not used.  Can you use this opportunity to remove it?

File: brkt_cli/__init__.py [hunk 6]
```diff
@@ -197,6 +157,147 @@ def main():
...
+    else:
+        zone = zones[0]
+        log.info('Using zone %s', zone)
+    encrypted_ami = values.ami
+    guest_ami_error = aws_svc.validate_guest_ami(encrypted_ami)
+    if guest_ami_error:
+        print(guest_ami_error, file=sys.stderr)
+        return 1
+    updater_ami = values.updater_ami
+    updater_ami_error = aws_svc.validate_encryptor_ami(values.updater_ami)
```
Comment (bburtin) [inline brkt_cli/__init__.py]: Is this the type of validation that we want to do?  Seems like we should be verifying that the encrypted AMI is one that we encrypted.

File: brkt_cli/__init__.py [hunk 7]
```diff
@@ -197,6 +157,136 @@ def main():
...
+    if updater_ami_error:
+        log.error('Update failed: %s', updater_ami_error)
+        return 1
+    # Initial validation done
+    log.info('Commencing update for AMI %s', encrypted_ami)
```
Comment (bburtin) [inline brkt_cli/__init__.py]: Nit: "Commencing update..." sounds wordy.  How about simply

log.info('Updating %', encrypted_ami)

Since the ID starts with "ami-", it's clear from the message that we're talking about an AMI.  Your call.

File: brkt_cli/update_encrypted_ami.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        encrypt_ami.NAME_METAVISOR_LOG_SNAPSHOT: mv_log_snapshot
+    }
+
+
+def retrieve_guest_volume_snapshot(aws_service,
+                                   encrypted_ami_id,
+                                   zone):
+    # Verify expected device mapping is present, return with info
+    log.info('Creating guest volume snapshot')
+    encrypted_image = aws_service.conn.get_image(encrypted_ami_id)
```
Comment (bburtin) [inline brkt_cli/update_encrypted_ami.py]: This should be aws_service.get_image(), to maintain the service abstraction.

File: brkt_cli/update_encrypted_ami.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    guest_volume_mapping = \
+        encrypted_image.block_device_mapping.get('/dev/sda5')
+    if not guest_volume_mapping:
+        return None, None, None, \
+            'Invalid block device mapping: /dev/sda5 not present'
+    snapshot = aws_service.conn.get_all_snapshots(
```
Comment (bburtin) [inline brkt_cli/update_encrypted_ami.py]: aws_service.get_snapshot()

File: brkt_cli/update_encrypted_ami.py (hunk 4 above)
Comment (bburtin) [inline brkt_cli/update_encrypted_ami.py]: I'm thinking the same thing as Dennis.  If we really need this logic, make sure we clean up resources that are created here.  From first glance, it looks like we could leak resources created inside this function if things go south.

File: brkt_cli/util.py [hunk 10]
```diff
@@ -38,3 +35,23 @@ def is_expired(self):
...
         """
         return self.clock.time() >= self.deadline
+
+
+def make_nonce():
+    """Returns a 32bit nonce in hex encoding"""
+    return str(uuid.uuid4()).split('-')[0]
+
+
+def validate_ami_name(name):
```
Comment (bburtin) [inline brkt_cli/util.py]: Unused?

File: brkt_cli/util.py [hunk 11]
```diff
@@ -38,3 +35,23 @@ def is_expired(self):
...
+    try:
+        service.validate_image_name(name)
+        return 0
+    except service.ImageNameError as e:
+        return 1
+
+
+def validate_region(region):
```
Comment (bburtin) [inline brkt_cli/util.py]: Use AWSService to get the list of regions.  Also, this function should return True/False, not 1/0.  Can this be in __init__.py, since it's the only callsite?

File: README.md (hunk 3 above)
Comment (rohang06) [inline README.md]: Fixed in next diff

File: brkt_cli/update_encrypted_ami.py (hunk 4 above)
Comment (rohang06) [inline brkt_cli/update_encrypted_ami.py]: I thought that if a customer had deregistered their AMI, the snapshot would go away. This isn't the case, so I'll remove this logic. Fixed in next diff

File: brkt_cli/__init__.py (hunk 5 above)
Comment (rohang06) [inline brkt_cli/__init__.py]: Fixed in next diff

File: brkt_cli/__init__.py (hunk 6 above)
Comment (rohang06) [inline brkt_cli/__init__.py]: I can't see a definable way to verify that this is the one we encrypted. For example:
image.description
u'Based on ami-76e27e1e, encrypted by Bracket Computing'
image.name
u'ubuntu/images/hvm/ubuntu-trusty-14.04-amd64-server-20141125 (encrypted 0c09d5a3)'
image.tags
{u'BrktEncryptorSessionID': u'1dcd61a0', u'BrktEncryptor': u'True', u'BrktEncryptorAMI': u'ami-c62259ac'}
Are all mutable so wouldn't prove anything. Moreover, shouldn't the user be keeping track of their bracket AMIs that they want to update?

File: brkt_cli/__init__.py (hunk 7 above)
Comment (rohang06) [inline brkt_cli/__init__.py]: fixed in next diff

File: brkt_cli/update_encrypted_ami.py (hunk 8 above)
Comment (rohang06) [inline brkt_cli/update_encrypted_ami.py]: fixed in next diff

File: brkt_cli/__init__.py (hunk 6 above)
Comment (bburtin) [inline brkt_cli/__init__.py]: Users can make mistakes, and we need to fail early and clearly when they do.  I'm pretty sure that we tag the encrypted AMI.  Can you check the tags during validation?

File: brkt_cli/update_encrypted_ami.py (hunk 9 above)
Comment (rohang06) [inline brkt_cli/update_encrypted_ami.py]: fixed in next diff

File: brkt_cli/util.py (hunk 10 above)
Comment (rohang06) [inline brkt_cli/util.py]: fixed in next diff

File: brkt_cli/util.py (hunk 11 above)
Comment (rohang06) [inline brkt_cli/util.py]: fixed in next diff

File: brkt_cli/update_encrypted_ami.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        encrypt_ami.NAME_METAVISOR_ROOT_SNAPSHOT: mv_root_snapshot,
+        encrypt_ami.NAME_METAVISOR_GRUB_SNAPSHOT: mv_grub_snapshot,
+        encrypt_ami.NAME_METAVISOR_LOG_SNAPSHOT: mv_log_snapshot
+    }
+
+
+def retrieve_guest_volume_snapshot(aws_service,
+                                   encrypted_ami_id):
+    # Verify expected device mapping is present, return with info
+    log.info('Creating guest volume snapshot')
```
Comment (bburtin) [inline brkt_cli/update_encrypted_ami.py]: This message is misleading.  Should be safe to omit it, since we're not modifying anything?

File: brkt_cli/update_encrypted_ami.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    if not guest_volume_mapping:
+        return None, None,\
+            'Invalid block device mapping: /dev/sda5 not present'
+    # Need to recover the volume iops and size
+    volume_info = {'iops': guest_volume_mapping.iops,
+                   'size': guest_volume_mapping.size,
+                   'type': guest_volume_mapping.volume_type}
+    guest_volume_snapshot = aws_service.get_snapshot(
+        guest_volume_mapping.snapshot_id)
+    return guest_volume_snapshot, volume_info, None
```
Comment (bburtin) [inline brkt_cli/update_encrypted_ami.py]: Get rid of the 3rd value?

File: brkt_cli/util.py [hunk 14]
```diff
@@ -19,11 +21,6 @@ class BracketError(Exception):
     pass
 
 
-def make_nonce():
```
Comment (bburtin) [inline brkt_cli/util.py]: Looks like you can revert the changes to this file.

**Comment by rohang06** [comment]
replaced by #78
